src/zones/erldns_zone_cache.erl

```diff
@@ -42,325 +27,249 @@
+            case ets:select_count(zone_records_typed, [Pattern]) of
+                0 ->
+                    is_name_in_zone_with_wildcard(NormalizedName, ZoneName);
                 _ ->
                     true
-            end;
-        _ ->
-            false
+            end
     end.
 
-%% @doc Return a list of tuples with each tuple as a name and the version SHA
-%% for the zone.
+-doc "Return a list of tuples with each tuple as a name and the version SHA for the zone.".
 -spec zone_names_and_versions() -> [{dns:dname(), binary()}].
 zone_names_and_versions() ->
-    % ETS and Mnesia foldl/3 differ in return values -> Mnesia's version is missing key
-    case erldns_config:storage_type() of
-        erldns_storage_json ->
-            erldns_storage:foldl(
-                fun({_, Zone}, NamesAndShas) ->
-                    NamesAndShas ++ [{Zone#zone.name, Zone#zone.version}]
-                end,
-                [],
-                zones
-            );
-        erldns_storage_mnesia ->
-            erldns_storage:foldl(
-                fun(Zone, NamesAndShas) ->
-                    NamesAndShas ++ [{Zone#zone.name, Zone#zone.version}]
-                end,
-                [],
-                zones
-            )
-    end.
+    ets:foldl(
+        fun(Zone, NamesAndShas) ->
+            [{Zone#zone.name, Zone#zone.version} | NamesAndShas]
+        end,
+        [],
+        zones
+    ).
 
-%% @doc Return current sync counter
+-doc "Return current sync counter".
 -spec get_rrset_sync_counter(dns:dname(), dns:dname(), dns:type()) -> integer().
 get_rrset_sync_counter(ZoneName, RRFqdn, Type) ->
-    case
-        erldns_storage:select(
-            sync_counters,
-            [
-                {{erldns:normalize_name(ZoneName), erldns:normalize_name(RRFqdn), Type, '$1'}, [], [
-                    '$_'
-                ]}
-            ],
-            infinite
-        )
-    of
-        [{ZoneName, RRFqdn, Type, Counter}] ->
-            Counter;
-        [] ->
-            % return default value of 0
-            0
-    end.
-
-%% @doc Update the RRSet sync counter for the given RR set name and type in the given zone.
--spec write_rrset_sync_counter({dns:dname(), dns:dname(), dns:type(), integer()}) -> ok.
-write_rrset_sync_counter({ZoneName, RRFqdn, Type, Counter}) ->
-    erldns_storage:insert(sync_counters, {ZoneName, RRFqdn, Type, Counter}).
+    NormalizedZoneName = dns:dname_to_lower(ZoneName),
+    NormalizedRRFqdn = dns:dname_to_lower(RRFqdn),
+    Key = {NormalizedZoneName, NormalizedRRFqdn, Type},
+    % return default value of 0
+    ets:lookup_element(sync_counters, Key, 2, 0).
+
+%% Update the RRSet sync counter for the given RR set name and type in the given zone.
+-spec write_rrset_sync_counter(dns:dname(), dns:dname(), dns:type(), integer()) -> ok.
+write_rrset_sync_counter(ZoneName, RRFqdn, Type, Counter) ->
+    true = ets:insert(sync_counters, {{ZoneName, RRFqdn, Type}, Counter}),
+    ok.
```

Comment:
We're doing the normalization on the read but not on the write.

Reply:
This is called only on two places with input that has already been normalised.